### PR TITLE
interact with CodeClimate to help automate repo health for sprint review

### PR DESCRIPTION
**Pull request by chessbyte**
Sample Output:

```sh
$ ruby code_climate.rb
Stats for Repo ManageIQ/manageiq: {"files"=>1742, "loc"=>83362, "issues"=>{"total"=>1068, "complexity"=>474, "duplication"=>293, "style"=>261, "bug_risk"=>40}, "rating"=>"A"}
Stats for Repo ManageIQ/manageiq-schema: {"files"=>848, "loc"=>17666, "issues"=>{"total"=>280, "complexity"=>68, "duplication"=>183, "style"=>27, "bug_risk"=>2}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-api: {"files"=>239, "loc"=>7729, "issues"=>{"total"=>194, "complexity"=>95, "duplication"=>90, "style"=>2, "bug_risk"=>7}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-ui-classic: {"files"=>2914, "loc"=>126440, "issues"=>{"total"=>4973, "complexity"=>2422, "duplication"=>976, "style"=>590, "bug_risk"=>219}, "rating"=>"C"}
Stats for Repo ManageIQ/manageiq-ui-service: {"files"=>408, "loc"=>13083, "issues"=>{"total"=>201, "complexity"=>177, "duplication"=>24, "style"=>0, "bug_risk"=>0}, "rating"=>"C"}
Stats for Repo ManageIQ/manageiq-automation_engine: {"files"=>204, "loc"=>8292, "issues"=>{"total"=>179, "complexity"=>147, "duplication"=>23, "style"=>7, "bug_risk"=>2}, "rating"=>"C"}
Stats for Repo ManageIQ/manageiq-content: {"files"=>1548, "loc"=>9210, "issues"=>{"total"=>248, "complexity"=>105, "duplication"=>107, "style"=>31, "bug_risk"=>5}, "rating"=>"C"}
Stats for Repo ManageIQ/manageiq-providers-amazon: {"files"=>125, "loc"=>4693, "issues"=>{"total"=>84, "complexity"=>52, "duplication"=>18, "style"=>13, "bug_risk"=>1}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-ansible_tower: {"files"=>75, "loc"=>1792, "issues"=>{"total"=>24, "complexity"=>15, "duplication"=>7, "style"=>1, "bug_risk"=>1}, "rating"=>"A"}
Stats for Repo ManageIQ/manageiq-providers-azure: {"files"=>101, "loc"=>3961, "issues"=>{"total"=>75, "complexity"=>51, "duplication"=>15, "style"=>5, "bug_risk"=>4}, "rating"=>"C"}
Stats for Repo ManageIQ/manageiq-providers-azure_stack: {}
Stats for Repo ManageIQ/manageiq-providers-foreman: {"files"=>56, "loc"=>805, "issues"=>{"total"=>11, "complexity"=>7, "duplication"=>2, "style"=>2, "bug_risk"=>0}, "rating"=>"A"}
Stats for Repo ManageIQ/manageiq-providers-google: {"files"=>84, "loc"=>1650, "issues"=>{"total"=>15, "complexity"=>10, "duplication"=>2, "style"=>3, "bug_risk"=>0}, "rating"=>"A"}
Stats for Repo ManageIQ/manageiq-providers-kubernetes: {"files"=>86, "loc"=>3886, "issues"=>{"total"=>49, "complexity"=>38, "duplication"=>5, "style"=>5, "bug_risk"=>1}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-lenovo: {"files"=>113, "loc"=>3671, "issues"=>{"total"=>52, "complexity"=>10, "duplication"=>38, "style"=>4, "bug_risk"=>0}, "rating"=>"A"}
Stats for Repo ManageIQ/manageiq-providers-nuage: {"files"=>83, "loc"=>1600, "issues"=>{"total"=>22, "complexity"=>8, "duplication"=>13, "style"=>1, "bug_risk"=>0}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-openshift: {"files"=>62, "loc"=>693, "issues"=>{"total"=>11, "complexity"=>5, "duplication"=>4, "style"=>1, "bug_risk"=>1}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-openstack: {"files"=>202, "loc"=>10390, "issues"=>{"total"=>227, "complexity"=>133, "duplication"=>79, "style"=>12, "bug_risk"=>3}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-ovirt: {"files"=>126, "loc"=>6621, "issues"=>{"total"=>102, "complexity"=>70, "duplication"=>7, "style"=>25, "bug_risk"=>0}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-redfish: {"files"=>69, "loc"=>1286, "issues"=>{"total"=>19, "complexity"=>6, "duplication"=>12, "style"=>1, "bug_risk"=>0}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-scvmm: {"files"=>67, "loc"=>1683, "issues"=>{"total"=>25, "complexity"=>14, "duplication"=>4, "style"=>7, "bug_risk"=>0}, "rating"=>"B"}
Stats for Repo ManageIQ/manageiq-providers-vmware: {"files"=>160, "loc"=>7735, "issues"=>{"total"=>208, "complexity"=>153, "duplication"=>26, "style"=>27, "bug_risk"=>2}, "rating"=>"C"}
```

### CODE REVIEW

File: code_climate.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
+require "net/http"
+require "json"
+require "active_support/core_ext"
+
+CODECLIMATE_API_TOKEN = "your CodeClimage API token"
```
Comment (Fryguy) [inline code_climate.rb:5]: Typo.  Even so, I would just enforce this as an ENV var...prevents changing the code and thus accidentally committing the code.

File: code_climate.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+CODECLIMATE_URL       = "https://api.codeclimate.com"
+
+class CodeClimateRequest
+  def initialize
+    @base_uri  = ENV["CODECLIMATE_URL"]       || CODECLIMATE_URL
+    @api_token = ENV["CODECLIMATE_API_TOKEN"] || CODECLIMATE_API_TOKEN
```
Comment (Fryguy) [inline code_climate.rb:11]: You can enforce it via code with `ENV.fetch("CODECLIMATE_API_TOKEN")`

File: code_climate.rb [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    snapshot = cc.snapshot(repo_id, snapshot_id)
+    files    = cc.files(repo_id, snapshot_id)
+    issues   = cc.issues(repo_id, snapshot_id)
+
+    issue_counts = {}
+    issue_counts['total']       = issues.length
+    issue_counts['complexity']  = issues.count { |issue| issue["attributes"]["categories"].include?("Complexity") }
+    issue_counts['duplication'] = issues.count { |issue| issue["attributes"]["categories"].include?("Duplication") }
+    issue_counts['style']       = issues.count { |issue| issue["attributes"]["categories"].include?("Style") }
+    issue_counts['bug_risk']    = issues.count { |issue| issue["attributes"]["categories"].include?("Bug Risk") }
```
Comment (Fryguy) [inline code_climate.rb:45]: Optional, but more_core_extensions has a helper for this called `element_counts`...Roughly something like the following I think would work...

```ruby
require "more_core_extensions/core_ext/array/element_counts"
issue_counts = issues.flat_map { |i| issue["attributes"]["categories"] }.element_counts
issue_counts['total'] = issues.length
```

File: code_climate.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
+require "net/http"
+require "json"
+require "active_support/core_ext"
+
+DEFAULT_CODECLIMATE_URL = "https://api.codeclimate.com"
+
+class CodeClimateRequest
+  def initialize
+    @base_uri  = ENV["CODECLIMATE_URL"] || DEFAULT_CODECLIMATE_URL
```
Comment (bdunne) [inline code_climate.rb:9]: We may want to use the config gem.  I think some of the other files in this repo started using it.

File: code_climate.rb [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      end
+    else
+      STDERR.puts "Invalid response code for path=\"#{path}\": #{response.code}"
+    end
+    
+    data
+  end
+end
+
+ALL_REPO_NAMES = [
```
Comment (bdunne) [inline code_climate.rb:131]: This could also live in the config

File: code_climate.rb [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  "ManageIQ/manageiq-providers-kubernetes",
+  "ManageIQ/manageiq-providers-lenovo",
+  "ManageIQ/manageiq-providers-nuage",
+  "ManageIQ/manageiq-providers-openshift",
+  "ManageIQ/manageiq-providers-openstack",
+  "ManageIQ/manageiq-providers-ovirt",
+  "ManageIQ/manageiq-providers-redfish",
+  "ManageIQ/manageiq-providers-scvmm",
+  "ManageIQ/manageiq-providers-vmware",
+]
```
Comment (Fryguy) [inline code_climate.rb:154]: This bit makes me think this should ultimately live in manageiq-release, but this is good for now :+1:

File: code_climate.rb (hunk 4 above)
Comment (bdunne) [reply code_climate.rb:9]: Nevermind, this is for later.  I don't see it in the Gemfile.  I guess scripts are using custom code :disappointed:

**Pull request merged**